lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -4168,18 +4172,29 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
       }
       domOrder.pushChildren(bb);
     }
-
+    
     // Create adjoint blocks and arguments, visiting original blocks in
     // post-order.
     for (auto *origBB : postOrderInfo->getPostOrder()) {
       auto *adjointBB = adjoint.createBasicBlock();
       adjointBBMap.insert({origBB, adjointBB});
-      // If adjoint block is the adjoint entry, continue.
-      if (adjointBB->isEntry())
-        continue;
-      // Otherwise, add a pullback struct argument to the adjoint block.
       auto pbStructLoweredType =
           remapType(getPullbackInfo().getPullbackStructLoweredType(origBB));
+      // If the BB is the original exit, then the adjoint block that we just
+      // createed must be the adjoint function's entry. We always know what an
+      // entry block's arguments should be, so we generate them and skip to the
+      // the next block.
+      if (origBB == origExit) {
+        assert(adjointBB->isEntry());
+        createEntryArguments(&getAdjoint());
+        auto *lastArg = adjointBB->getArguments().back();
+        assert(lastArg->getType() == pbStructLoweredType);
+        adjointPullbackStructArguments[origBB] = lastArg;
+        continue;
+      }
+      
+      // Otherwise, we create a phi argument for the corresponding primal value
```

Comment:
I'd suggest:
```suggestion
      // Otherwise, create a phi argument for the pullback struct and handle
      // dominated active values/buffers.
```

"Primal value struct" has been renamed to "pullback struct".
Comments below describe how exactly dominated active values/buffers are handled.